File: 2016/day06/main.hs
import Data.List (group, sort, sortOn, transpose)
import Control.Monad (forM_)

columnValuesByFrequencyAscending :: (Ord a, Eq a) => [[a]] -> [[a]]
columnValuesByFrequencyAscending =
  map ( map head
      . sortOn length
      . group
      . sort
      )
  . transpose

main :: IO()
main = do
  result <- columnValuesByFrequencyAscending . lines <$> readFile "input.txt"
  forM_ [("One", last), ("Two", head)] $ \(part, aggregator) ->
    putStrLn $ "Part " ++ part ++ ": " ++ map aggregator result

File: 2016/day06/main.rb
require_relative "../day04/frequencies"

signals = IO.readlines("input.txt").map do |line|
  line.strip.chars
end

column_frequencies = signals.transpose.map(&method(:frequencies))

column_values_by_frequency_ascending = column_frequencies.map do |frequencies|
 frequencies.to_a.sort_by(&:reverse).map(&:first)
end

parts = {"One" => :last, "Two" => :first}

parts.each do |part, aggregator|
  result = column_values_by_frequency_ascending.map(&aggregator)
  puts "Part #{part}: #{result.join}"
end
